src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs

```diff
@@ -13,11 +14,14 @@ namespace Microsoft.Extensions.Primitives
     /// </summary>
     public class CompositeChangeToken : IChangeToken
     {
-        private static readonly Action<object> _onChangeDelegate = OnChange;
-        private readonly object _callbackLock = new object();
-        private CancellationTokenSource _cancellationTokenSource;
-        private bool _registeredCallbackProxy;
-        private List<IDisposable> _disposables;
+        private static readonly Action<object?> _onChangeDelegate = OnChange;
+        private readonly object _callbackLock = new();
+        private CancellationTokenSource? _cancellationTokenSource;
+        private List<IDisposable>? _disposables;
+
+        [MemberNotNullWhen(true, nameof(_cancellationTokenSource))]
+        [MemberNotNullWhen(true, nameof(_disposables))]
```

Comment:
I don't believe this will make any difference for the compiler (at least it doesn't on a test app I just tried) - also you're not doing anything to guarantee this in here. cc: @stephentoub (in case I'm wrong or there is something new coming)

Reply:
actually I tried a bit different pattern than you have here and that one didn't work but specific case you have here works...

Reply:
It's needed for `EnsureCallbacksInitialized` method to work. Since we returning from it if `_registeredCallbackProxy` is `true`, compiler complains that nothing is initialized.

And because `_registeredCallbackProxy` is set to `true` only an the end of that method, it can be used as a sign that those fields are initialized.

Reply:
https://github.com/dotnet/roslyn/issues/55664

Reply:
yes, noticed that, the behavior I described seems to be a bug I referenced above